### PR TITLE
remove unused `type: ignore` directives

### PR DESCRIPTION
**Pull request by pmeier**
During development it is common practice to put `type: ignore` comments on lines that are correct, but `mypy` doesn't recognize this. This often stems from the fact, that the used `mypy` version wasn't able to handle the used pattern.

With every new release `mypy` gets better at handling complex code. In addition to fix all the previously accepted but now failing patterns, we should also revisit all `type: ignore` comments to see if they are still needed or not. Fortunately, we don't need to do it manually: by adding `warn_unused_ignores = True` to the configuration, `mypy` will error out in case it encounters an `type: ignore` that is no longer needed.

### CODE REVIEW

File: torch/nn/utils/parametrizations.py [hunk 1]
```diff
@@ -90,7 +90,7 @@ def _power_method(self, weight_mat: torch.Tensor, n_power_iterations: int) -> No
             # This power iteration produces approximations of `u` and `v`.
             self._u = F.normalize(torch.mv(weight_mat, self._v),      # type: ignore[has-type]
                                   dim=0, eps=self.eps, out=self._u)   # type: ignore[has-type]
-            self._v = F.normalize(torch.mv(weight_mat.t(), self._u),  # type: ignore[has-type]
+            self._v = F.normalize(torch.mv(weight_mat.t(), self._u),
```
Comment (albanD) [inline torch/nn/utils/parametrizations.py:93]: Can you do the same for the line above?

**Review by albanD — APPROVED** [review]
LGTM

**Review by rgommers — APPROVED** [review]
LGTM too, this is useful. Thanks @pmeier

File: torch/nn/utils/parametrizations.py (hunk 1 above)
Comment (pmeier) [reply torch/nn/utils/parametrizations.py:93]: This was not flagged by `mypy` as unused so I kept it.

**Comment by albanD** [comment]
Thiis is going to be tricky to land without race. I'll wait for the week end to do that :D

**Comment by albanD** [comment]
Early morning rebase/land worked :D 
Sorry for the force push into your branch, I had to after the rebase.

**Comment by samestep** [comment]
@pmeier I wish you had added me as a reviewer on this PR 😅 I'm glad about removing most of the unused `type: ignore` comments that this PR removes, but a couple of them were there for a specific reason. Example: not everyone has [`ruamel.yaml`](https://pypi.org/project/ruamel.yaml/) installed, and after this PR, `mypy` only passes if you _don't_ have it installed. If you do have it installed (as I do on my machine), running `mypy --config mypy-strict.ini` fails with this output:
```
.github/scripts/lint_native_functions.py:17: error: unused 'type: ignore' comment
.github/scripts/lint_native_functions.py:29:8: error: Module has no attribute "YAML"  [attr-defined]
Found 2 errors in 1 file (checked 142 source files)
```
This is why I specifically _removed_ `warn_unused_ignores = True` from `mypy-strict.ini` in #59117, citing https://github.com/pytorch/pytorch/pull/56402#issuecomment-822743795:

> Possible downsides of doing this:
>
> 1. This could cause un-ignorable errors on files that are checked by _both_ `mypy.ini` and `mypy-strict.ini`.
> 2. This may cause `mypy` errors that appear on some platforms or environments but not on others.

In my opinion, while enabling this option has benefits (as shown by this PR), it makes `mypy` too brittle as a result of being too sensitive to the set of packages that individual developers have (or do not have) installed.

I think there are a couple possible solutions to this:

1. Install all dev packages in the Lint / mypy job, fix the `type: ignore` comments in `.github/scripts/lint_native_functions.py`, and just be OK with the fact that devs will need to install `ruamel.yaml` if they want `mypy` to pass locally.
2. Explicitly set `warn_unused_ignores = False` in `mypy*.ini`, and add a comment (perhaps linking to this discussion) explaining why.

Thoughts? (It would also be good to hear from the rest of the @pytorch/pytorch-dev-infra team on this.)

**Comment by samestep** [comment]
It looks like this is currently blocking #60472.

**Comment by rgommers** [comment]
Hmm this makes sense, unfortunately. I'm in favor of (2), because I think Mypy should pass with only the default build-time and runtime dependencies installed.

**Comment by bluenote10** [comment]
In the meantime mypy has received a new feature that would allow for a better solution: https://github.com/python/mypy/issues/8823#issuecomment-1556288015

It should now be possible to use e.g. `# type: ignore[import-not-found, unused-ignore]` for the example with the ruamel import. The idea is basically than one can now combined `unused-ignore` code with any other code to express type errors the _may or may not be there_. Here is a [demonstration of that on mypy playground](https://mypy-play.net/?mypy=latest&python=3.12&flags=warn-unused-ignores&gist=2aec3e18bcd997eee762289ef84a82d1).

This new feature would allow to enable `warn_unused_ignores = True` globally, without causing trouble for these "maybe a type error" cases. @samestep Given this new feature, would you be fine with re-enabling unused ignore detection?

Clearly, having "dead" type ignores lingering around is pretty dangerous, because while they may not hide a bug today, they could suppress a genuine bug at any time in the future. In the worst case, one could sprinkle half the code base with type ignores, making the whole idea of using a type checker pointless.

**Comment by samestep** [comment]
@bluenote10 Oh awesome! Yeah that sounds perfect. I don't work at Meta anymore though, so I have no authority here. @malfet do you know who would be the right person to look at this?

**Comment by malfet** [comment]
@samestep anyone who is willing to submit the PR. :) And I'll be happy to review it. mypy version is defined in lintrunner.yaml

**Comment by bluenote10** [comment]
> anyone who is willing to submit the PR

Great, I can give it a shot, if I find some time in the next days / weeks.

**Pull request closed without merging**